### PR TITLE
Improvement: Optimize # of RPC roundtrips to reconcile log inconsistencies

### PR DESCRIPTION
**Pull request by faineance**
> If desired, the protocol can be optimized to reduce the number of rejected AppendEntries RPCs. For example, when rejecting an AppendEntries request, the follower can include the term of the conflicting entry and the first index it stores for that term. With this information, the leader can decrement nextIndex to bypass all of the conflicting entries in that term; one AppendEntries RPC will be required for each term with conflicting entries, rather than one RPC per entry

It's not clear from the text exactly how the conflicting index and term sent back from the client should be used by the leader to determine what nextIndex to use. I believe the protocol the authors intended is as follows:
- If a follower does not have prevLogIndex in its log, it should return the latest log index it has, and no term.
- If a follower does have prevLogIndex in its log, but the term does not match, it should return the term of the conflicting entry, and then find the first entry in that term.
- Upon receiving a conflict response, the leader should first search its log for conflicted term. If it finds an entry in its log with that term, it should set nextIndex to be the one beyond the index of the last entry in that term in its log. If it does not find an entry with that term, it should set nextIndex to the conflicted index. 

A less optimal version is just to have the follower send back the index. ( this'd involve the leader sending more log entries then necessary to bring them up to date )

Implementation changes:
-

### CODE REVIEW

File: src/Raft/RPC.hs [hunk 1]
```diff
@@ -103,12 +103,27 @@ data AppendEntries v = AppendEntries
...
 
+-- | To reduce the number of rejected AppendEntries RPCs when inconsistencies between
+-- leader and followers' logs occur, the follower includes the term of the
+-- conflicting entry and the first index it stores for that term.
+-- The leader can then decrement the nextIndex accordingly and bypass all the
+-- conflicting entries.
+data AppendEntriesResponseStatus
+  = AERSuccess
+  | AERStaleTerm -- ^ term given was stale (older) then followers term
+  | AERConflict {
```
Comment (faineance) [inline src/Raft/RPC.hs]: Maybe AERInconsistency would be a better name seeing as this is also returned when the follower is more then one log entry out of date?

Comment (tdietert) [reply src/Raft/RPC.hs]: :+1:

File: src/Raft/Follower.hs [hunk 2]
```diff
@@ -40,44 +40,59 @@ import Raft.Types
...
+      --
+      case fsTermAtAEPrevIndex fs of
+        Nothing
+          -- there are no previous entries
+          | aePrevLogIndex == index0 -> do
+              appendLogEntries aeEntries
+              pure (AERSuccess, updateFollowerState fs)
+          -- the follower doesn't have the previous index given in the AppendEntriesRPC
+          | otherwise -> do
+              let conflict = AERConflict {
```
Comment (faineance) [inline src/Raft/Follower.hs]: It'd be nice to not have duplicated the creation of the same AERConflict. 
Matching with pattern guards on the Nothing + other conditions seems dirty, and I'm not sure about putting it conflict in the where or above.

File: src/Raft/Leader.hs [hunk 3]
```diff
@@ -45,17 +45,20 @@ handleAppendEntries (NodeLeaderState ls)_ _  =
...
 
 handleAppendEntriesResponse :: forall sm v. RPCHandler 'Leader sm AppendEntriesResponse v
-handleAppendEntriesResponse ns@(NodeLeaderState ls) sender appendEntriesResp
-  -- If AppendEntries fails (aerSuccess == False) because of log inconsistency,
-  -- decrement nextIndex and retry
-  | not (aerSuccess appendEntriesResp) = do
-      let newNextIndices = Map.adjust decrIndexWithDefault0 sender (lsNextIndex ls)
+handleAppendEntriesResponse ns@(NodeLeaderState ls) sender appendEntriesResp =
+  case aerStatus appendEntriesResp of
+    -- If a follower receivers an AppendEntries with a stale term number, it
```
Comment (tdietert) [inline src/Raft/Leader.hs]: it... what?! :fearful:

File: src/Raft/Leader.hs [hunk 4]
```diff
@@ -45,17 +45,20 @@ handleAppendEntries (NodeLeaderState ls)_ _  =
...
+    AERStaleTerm -> pure (leaderResultState Noop ls)
+    -- If AppendEntries fails because of log inconsistency,
+    -- decrement the nextIndex to the first index of the conflicting term and retry
+    AERInconsistent aerFirstIndexStoredForTerm -> do
+      let newNextIndices = Map.insert sender aerFirstIndexStoredForTerm (lsNextIndex ls)
```
Comment (tdietert) [inline src/Raft/Leader.hs]: :ok_hand: this is the money

File: src/Raft/Follower.hs [hunk 5]
```diff
@@ -40,43 +41,27 @@ import Raft.Types
...
-                  appendLogEntries aeEntries
-                  -- 5. If leaderCommit > commitIndex, set commitIndex =
-                  -- min(leaderCommit, index of last new entry)
-                  pure (True, updateFollowerState fs)
-    when success resetElectionTimeout
+
+    let status = handleAppendEntries' currentTerm fs ae
+
+    newFollowerState <-
+      if status == AERSuccess
```
Comment (tdietert) [inline src/Raft/Follower.hs]: a small nitpick here, but I'd prefer a case statement:

```suggestion
    case handleAppendEntries' currentTerm fs ae of
      AERSuccess -> ...
      AERStaleTerm -> pure fs
      AERInconsistent -> pure fs
```

Comment (tdietert) [reply src/Raft/Follower.hs]: This is not a verbatim "github suggestion", just sketching out the case statement.

**Comment by faineance** [comment]
On pause till tests pass consistently on master

**Comment by faineance** [comment]
closing in favour of https://github.com/adjoint-io/raft/pull/130
too many merge conflicts here

**Pull request closed without merging**